### DISCUSSION TITLE
Sending files via sound, js example

### DISCUSSION THREAD

**Discussion by yaroslav-perec** [ggerganov/ggwave#63] [Q&A] [answered]
Hi! First of all, thank you for the great library!

I've found that using waver we can send files from one device to another. Can this be done inside a web application? If so, are there any examples of how to do it correctly?

**Comment by ggerganov** [comment, marked as answer]
Hi, thanks for the feedback!

The web application does not support sharing files.
The reason is that Waver uses the TCP protocol to perform the actual file transfer and web browsers do not allow peer-to-peer TCP connections.

Note that the mobile application (iOS and Android) uses sound only to establish the connection between the devices. The actual data transfer is performed over the TCP protocol.

You might want to check the [wave-share](https://github.com/ggerganov/wave-share) project which allows web applications to share files with each other using data-over-sound. Note that only the **signalling** is performed via sound. The actual data transfer is performed via WebRTC.

**Reply by yaroslav-perec** [reply]
Thanks for the detailed answer! 
One more question, could you please tell me if the following use case is possible: two devices, one is a TV with speakers but without a microphone, the second device is a smartphone that has both a microphone and a speaker, is it possible to somehow via sound establish a connection between these two devices and then send some data from a smartphone to a TV (though TCP or WebRTC) or both devices must have a microphone to establish a connection?

**Reply by ggerganov** [reply]
In theory it should be possible. That is how Waver's file sharing works.

For example, the TV can broadcast it's IP and port and start listening for all incoming connections. If a smartphone captures this information via microphone, it can try to connect to the TV. I think the connection has to be TCP, because for WebRTC you need a 2-way communication and without a mic on the TV, you don't have it. Maybe WebRTC + STUN/TURN can work, but I don't know the details.

For the TCP option, you would need app running on the TV that can open and listen on a TCP/IP port - I don't know if that is possible.
For the WebRTC + STUN/TURN, you can use the TV's web browser.

**Reply by yaroslav-perec** [reply]
Many thanks! I will try to investigate the option with WebRTC + STUN/TURN
